pkg/integrations/v2/app_o11y_receiver/app_o11y_receiver.go

```diff
@@ -0,0 +1,241 @@
+		instanceID:      id,
+		common:          c.Common,
+		globals:         globals,
+		conf:            c.ReceiverConfig,
+		handler:         handler,
+		exporters:       exp,
+		reg:             reg,
+	}, nil
+}
+
+// Handler is the http endpoint for exposing Prometheus metrics
+func (i *appo11yIntegration) Handler(prefix string) (http.Handler, error) {
+	r := mux.NewRouter()
+	r.Handle(path.Join(prefix, "metrics"), promhttp.HandlerFor(i.reg, promhttp.HandlerOpts{}))
+	return r, nil
+}
+
+// Targets implements MetricsIntegration
+func (i *appo11yIntegration) Targets(ep integrations.Endpoint) []*targetgroup.Group {
+	integrationNameValue := model.LabelValue("integrations/" + i.integrationName)
+
+	group := &targetgroup.Group{
+		Labels: model.LabelSet{
+			model.InstanceLabel: model.LabelValue(i.instanceID),
+			model.JobLabel:      integrationNameValue,
+			"agent_hostname":    model.LabelValue(i.globals.AgentIdentifier),
+
+			// Meta labels that can be used during SD.
+			"__meta_agent_integration_name":     model.LabelValue(i.integrationName),
+			"__meta_agent_integration_instance": model.LabelValue(i.instanceID),
+		},
+		Source: fmt.Sprintf("%s/%s", i.integrationName, i.instanceID),
+	}
+
+	group.Targets = append(group.Targets, model.LabelSet{
+		model.AddressLabel:     model.LabelValue(ep.Host),
+		model.MetricsPathLabel: model.LabelValue(path.Join(ep.Prefix, "/metrics")),
+	})
+
+	return []*targetgroup.Group{group}
+}
+
+// ScrapeConfigs implements MetricsIntegration
+func (i *appo11yIntegration) ScrapeConfigs(sd discovery.Configs) []*autoscrape.ScrapeConfig {
+	if !*i.common.Autoscrape.Enable {
+		return nil
+	}
+
+	cfg := promConfig.DefaultScrapeConfig
+	cfg.JobName = fmt.Sprintf("%s/%s", i.integrationName, i.instanceID)
+	cfg.Scheme = i.globals.AgentBaseURL.Scheme
+	cfg.HTTPClientConfig = i.globals.SubsystemOpts.ClientConfig
+	cfg.ServiceDiscoveryConfigs = sd
+	cfg.ScrapeInterval = i.common.Autoscrape.ScrapeInterval
+	cfg.ScrapeTimeout = i.common.Autoscrape.ScrapeTimeout
+	cfg.RelabelConfigs = i.common.Autoscrape.RelabelConfigs
+	cfg.MetricRelabelConfigs = i.common.Autoscrape.MetricRelabelConfigs
+
+	return []*autoscrape.ScrapeConfig{{
+		Instance: i.common.Autoscrape.MetricsInstance,
+		Config:   cfg,
+	}}
+}
+
+// RunIntegration implements Integration
+func (i *appo11yIntegration) RunIntegration(ctx context.Context) error {
+	mdlw := middleware.New(middleware.Config{
+		Recorder: metrics.NewRecorder(metrics.Config{Registry: i.reg, Prefix: utils.MetricsNamespace}),
+	})
+
+	r := mux.NewRouter()
+	r.Handle("/collect", i.handler.HTTPHandler(i.logger))
+
+	srv := &http.Server{
+		Addr:    fmt.Sprintf("%s:%d", i.conf.Server.Host, i.conf.Server.Port),
+		Handler: std.Handler("", mdlw, r),
+	}
+	errChan := make(chan error)
+
+	go func() {
+		level.Info(i.logger).Log("msg", "starting app o11y receiver", "host", i.conf.Server.Host, "port", i.conf.Server.Port)
+		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+			errChan <- err
```

Comment:
This may block forever if the integration exits before an error gets written; use a buffered channel for errChan to avoid that.

Reply:
- [ ] use buffered channel